core/platform/email/gae_email_services_test.py

```diff
@@ -38,8 +44,155 @@ def test_sending_email(self):
+            gae_email_services.send_mail(
+                malformed_sender_email, self.RECIPIENT_EMAIL,
+                'subject', 'body', 'html')
+        messages = self.mail_stub.get_sent_messages(
+            to=self.RECIPIENT_EMAIL)
+        self.assertEqual(0, len(messages))
+
+    def test_that_email_not_sent_if_recipient_address_is_malformed(self):
+        malformed_recipient_email = ''
+        email_exception = (self.assertRaisesRegexp(
+            ValueError, 'Malformed recipient email address: %s'
+            % malformed_recipient_email))
+        with self.swap(feconf, 'CAN_SEND_EMAILS', True), email_exception:
+            gae_email_services.send_mail(
+                feconf.ADMIN_EMAIL_ADDRESS, malformed_recipient_email,
+                'subject', 'body', 'html')
+
+
+class BulkEmailsTests(test_utils.GenericTestBase):
+    SENDER_EMAIL = 'sender@example.com'
+    SENDER_USERNAME = 'sender'
+    RECIPIENT_A_EMAIL = 'a@example.com'
+    RECIPIENT_A_USERNAME = 'usera'
+    RECIPIENT_B_EMAIL = 'b@example.com'
+    RECIPIENT_B_USERNAME = 'userb'
+    RECIPIENT_EMAILS = [RECIPIENT_A_EMAIL, RECIPIENT_B_EMAIL]
+
+    def setUp(self):
+        super(BulkEmailsTests, self).setUp()
+        # SENDER is authorised sender.
+        # A and B are recipients.
+        self.signup(self.SENDER_EMAIL, self.SENDER_USERNAME)
+        self.signup(self.RECIPIENT_A_EMAIL, self.RECIPIENT_A_USERNAME)
+        self.signup(self.RECIPIENT_B_EMAIL, self.RECIPIENT_B_USERNAME)
+        self.recipient_emails = [self.RECIPIENT_A_EMAIL, self.RECIPIENT_B_EMAIL]
+
+    def test_that_correct_bulk_emails_sent(self):
+        with self.swap(feconf, 'CAN_SEND_EMAILS', True):
+            gae_email_services.send_bulk_mail(
+                self.SENDER_EMAIL, self.RECIPIENT_EMAILS,
+                'subject', 'body', 'html')
+        message_a = self.mail_stub.get_sent_messages(
+            to=self.RECIPIENT_EMAILS[0])
+        self.assertEqual(len(message_a), 1)
+        message_b = self.mail_stub.get_sent_messages(
+            to=self.RECIPIENT_EMAILS[1])
+        self.assertEqual(len(message_b), 1)
+
+    def test_that_bulk_emails_not_sent_if_can_send_emails_is_false(self):
+        # Emails are not sent if the CAN_SEND_EMAILS setting is not turned on.
+        email_exception = (self.assertRaisesRegexp(
+            Exception, 'This app cannot send emails.'))
+        with email_exception:
+            gae_email_services.send_bulk_mail(
+                self.SENDER_EMAIL, self.recipient_emails,
+                'subject', 'body', 'html')
+        message_a = self.mail_stub.get_sent_messages(
+            to=self.RECIPIENT_EMAILS[0])
+        self.assertEqual(len(message_a), 0)
+
+        message_b = self.mail_stub.get_sent_messages(
+            to=self.RECIPIENT_EMAILS[1])
+        self.assertEqual(len(message_b), 0)
+
+    def test_that_bulk_mails_not_sent_if_sender_email_is_malformed(self):
+        malformed_sender_email = ''
```

Comment:
Can we make this test follow this pattern:
- Try sending bulk email with the correct sender email
- check for expected behavior
- Try sending bulk email with invalid sender email
- ensure it raises an error!  

(ditto in other places, sorry I missed this in the last review.)

Reply:
Bulk mail with correct sender_email is tested using:
`def test_that_correct_bulk_emails_sent(self):`
(Line Number 129)

The test for false sender_email is separate. We can combine those two though. This will increase the length of these tests. I am not sure what I should do with this now.
Looking forward to your suggestions.
Thanks!

Reply:
@DubeySandeep 
Any suggestions?
Thanks!

Reply:
@siddhantkhandelwal, through  `test_that_correct_bulk_emails_sent` you are testing the overall functionality, whereas through `test_that_bulk_mails_not_sent_if_sender_email_is_malformed` we are checking that an exception is raised under a certain condition, I was asking you to have a test to ensure that it doesn't raise any exception when email is correct. 

Also, I realized that all of the tests depends on correct email so I think we can leave it!

Reply:
Thanks!